applications/Preprocessor/lib/gmsh.cpp

```diff
@@ -0,0 +1,446 @@
+        });
+
+    if (two_dim) {
+        dimension--;
+        bool one_dim = std::all_of(
+            nodes_.begin(), nodes_.end(), [&](const MeshSource2::Coord& node) {
+                return (std::abs(node.coordinate[1]) < tol);
+            });
+        if (one_dim) {
+            dimension--;
+        }
+    }
+    logger(VERBOSE, "Dimension of the net is %", dimension);
+    return dimension;
+}
+
+void GmshReader::readElementData() {
+
+    bool found_elements = locate_in_file(Filehandle_, "$ElementData");
+    logger.assert_always(
+        found_elements,
+        "ElementData section of gmsh file  not found. HPGem needs this");
+    size_t numStringelements;
+    Filehandle_ >> numStringelements;
+    for (size_t i = 0; i < numStringelements; i++) {
+        std::string view_names;
+        Filehandle_ >> view_names;
+    }
+    size_t numrealelements;
+    Filehandle_ >> numrealelements;
+    for (size_t i = 0; i < numrealelements; i++) {
+        double real;
+        Filehandle_ >> real;
+    }
+    size_t numintegerelements;
+    Filehandle_ >> numintegerelements;
+    logger.assert_always(numintegerelements > 2,
+                         "There is no Elementspecific data in gmsh file");
+    size_t timestep;
+    size_t field_components;
+    size_t num_enteties;
+    Filehandle_ >> timestep >> field_components >> num_enteties;
+    for (int i = 3; i < numintegerelements; i++) {
+        size_t dummy;
+        Filehandle_ >> dummy;
+    }
+    for (int i = 0; i < num_enteties; i++) {
+        size_t elementid;
+        size_t zoneinfo;
+        Filehandle_ >> elementid >> zoneinfo;
+        elementid--;  // gmsh is 1 indexed hpgem is zero indexed
+        logger.assert_always(elements_[elementid].id == elementid,
+                             "Your elements are not indexed continously. This "
+                             "parser cannot handle it. % vs %",
+                             elements_[elementid].id, elementid);
+        elements_[elementid].zoneName = std::to_string(zoneinfo);
+    }
+}
+
+void GmshReader::readPBCs() {
+
+    bool found_PBCs = locate_in_file(Filehandle_, "$Periodic");
+    if (!found_PBCs) {
+        return;
+    }
+
+    size_t num_periodic_elements;
+    Filehandle_ >> num_periodic_elements;
+
+    for (size_t i = 0; i < num_periodic_elements; i++) {
+        size_t dummy;
+        Filehandle_ >> dummy;
+    }
+
+    std::unordered_map<size_t, size_t> pbc_nodes;
+
+    size_t periodic_notes;
+    Filehandle_ >> periodic_notes;
+    logger(VERBOSE, "Found % periodic nodes", periodic_notes);
+    for (size_t i = 0; i < periodic_notes; i++) {
+        size_t node1;
+        size_t node2;
+        Filehandle_ >> node2 >> node1;
+        // gmsh is 1 indexed hpgem is zero indexed
+        pbc_nodes[node2 - 1] = node1 - 1;
+    }
+
+    // because element ids are indices into the coordinate vector for the pbcs
+    // all we have to do is renumber the node ids.
+    for (auto& node : nodes_) {
```

Comment:
Is this going to work when there is more than one node that is linked up? For example, consider a square domain with periodic boundary conditions linking the top to bottom and a second one linking left to right. Then the four corner nodes will collapse to a single node. Looking at the code I'm very not sure whether this actually renumbers them all to a single nodeid.

Reply:
I am not a 100% sure on that but, gmsh has a concept of master nodes. I think this should work. But I will check the error condition